### PR TITLE
chore!: split `ValidationError` from `MaxSizeError`; `validate` and `max_serialized_size` made `BorshSchemaContainer`'s methods

### PR DESCRIPTION
**Pull request by dj8yfo**
split 
```rust
pub enum MaxSizeError {
    Overflow,
    Recursive,
    MissingDefinition,
    ZSTSequenceNotArray,
}
```
into 
```rust
pub enum MaxSizeError {
    Overflow,
    Recursive,
    MissingDefinition,
}
```
and 
```rust
pub enum ValidateError {
    /// sequences of zero sized types of dynamic length are forbidden by definition
    /// see <https://github.com/near/borsh-rs/pull/202> and related ones
    ZSTSequence,
}
```

### CODE REVIEW

**Comment by dj8yfo** [comment]
a valid alternative might be to remove second commit and newly added module `borsh/src/schema/container_ext/validate.rs`, 
and just remove `MaxSizeError::ZSTSequenceNotArray` variant and corresponding check

File: borsh/src/schema/container_ext/max_size.rs [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+    #[test]
+    fn test_is_zero_size_recursive_check_err() {
+        use crate as borsh;
+
+        #[derive(::borsh_derive::BorshSchema)]
+        struct RecursiveNoExitStructUnnamed(Box<RecursiveNoExitStructUnnamed>);
+
+        let schema = BorshSchemaContainer::for_type::<RecursiveNoExitStructUnnamed>();
+        assert_eq!(Err(()), is_zero_size(schema.declaration(), &schema));
+    }
```
Comment (dj8yfo) [inline borsh/src/schema/container_ext/max_size.rs:308]: a similar test in current master was causing stack overflow
```rust
    #[test]
    fn test_is_zero_size_recursive_check_err() {
        use crate as borsh;

        #[derive(::borsh_derive::BorshSchema)]
        struct RecursiveNoExitStructUnnamed(Box<RecursiveNoExitStructUnnamed>);

        let schema = BorshSchemaContainer::for_type::<RecursiveNoExitStructUnnamed>();
        assert_eq!(false, is_zero_size(schema.declaration(), &schema));
    }
```
```
thread 'schema_helpers::tests::test_is_zero_size_recursive_check_err' has overflowed its stack
fatal runtime error: stack overflow
```
Though rustc prevents from constructing a type, which would hit a [branch with recursing indefinetely into `is_zero_size`](https://github.com/near/borsh-rs/blob/master/borsh/src/schema_helpers.rs#L189):

```bash
error: values of the type `[[[RecursiveNoExitStructUnnamed2; 4294967295]; 2]; 4294967295]` are too big for the current architecture
```

File: borsh/src/schema/container_ext/validate.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    /// specifying `"nil"` as their [Declaration] for this method to work correctly.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use borsh::schema::BorshSchemaContainer;
+    ///
+    /// let schema = BorshSchemaContainer::for_type::<usize>();
+    /// assert_eq!(Ok(()), schema.validate());
+    ///
```
Comment (frol) [inline borsh/src/schema/container_ext/validate.rs:19]:
```suggestion
    /// ```
```

**Review by frol — APPROVED** [review]
Looks good to me 👍

**Pull request merged**